src/lib.rs

```diff
@@ -107,5 +94,47 @@
 //! - When defining a generic struct, you cannot use `VALUE` as a generic
 //!   parameter as this is what all setters are using.
 
-mod parse_struct;
-mod derive_builder;
+#![crate_type = "proc-macro"]
+#![feature(proc_macro, proc_macro_lib)]
+
+extern crate proc_macro;
+extern crate syn;
+#[macro_use]
+extern crate quote;
+
+use proc_macro::TokenStream;
+
+#[proc_macro_derive(Builder)]
+pub fn derive(input: TokenStream) -> TokenStream {
+    let input: String = input.to_string();
+
+    let ast = syn::parse_macro_input(&input).expect("Couldn't parse item");
+
+    let result = builder_for_struct(ast);
+
+    format!("{}\n{}", input, result).parse().expect("couldn't parse string to tokens")
+}
+
+fn builder_for_struct(ast: syn::MacroInput) -> quote::Tokens {
+    let fields = match ast.body {
+        syn::Body::Struct(syn::VariantData::Struct(ref fields)) => fields,
+        _ => panic!("#[derive(new)] can only be used with braced structs"),
+    };
+
+    let name = &ast.ident;
+    let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
+    let funcs = fields.iter().map(|f| {
+        let f_name = &f.ident;
+        let ty = &f.ty;
+        quote!(pub fn #f_name<VALUE: Into<#ty>>(&mut self, value: VALUE) -> &mut Self {
+            self.#f_name = value.into();
+            self
+        })
+    });
+
+    quote! {
+        impl #impl_generics #name #ty_generics #where_clause {
```

Comment:
Technical details which in no way affect my preference about where to put spaces:

- From the `quote!` macro's perspective there is no way for it to tell where you put spaces. It sees a `Vec<TokenTree>` which looks like `["impl", "#", "impl_generics", "#", "name", "#", "ty_generics", "#", "where_clause", Braces(["#", Parens(["funcs"]), "*"])]`.
- The string constructed by the `quote!` macro will have a single space in between every token (at least as of quote 0.3.3; I think in this version there might be more than one in certain cases). This is the simplest heuristic that is always correct with respect to rustc being able to parse the result as intended. So if `impl_generics` is `<T>` then the string will start with `"impl < T >".
- The builtin `stringify!` has the same behavior for the same reason:

    ```rust
    fn main() {
        // impl # impl_generics # name { # ( funcs ) * }
        println!("{}", stringify!(impl #impl_generics #name { #(funcs)* }));
    }
    ```

- When you take the string from `quote!` and parse it to a TokenStream for #[derive(Builder)] to return, the spaces are throw out again. The TokenStream looks like `["impl", "<", "T", ">", ...]`.
- One of the first things Rust does upon getting back the TokenStream is parse it. So if you [`cargo expand`](https://github.com/dtolnay/cargo-expand) a crate containing #[derive(Builder)] the spaces will be where Rust's pretty-printer puts them, not where `quote!` puts them and not where you put them.

Nontechnical details which fully determine my preference:

- I find `impl#impl_generics #name#ty_generics` less readable.
- I prefer the style @badboy used here.
- He probably got it from my examples so it is no coincidence :stuck_out_tongue:.

Bottom line:

- Put spaces wherever you want.
- There are so many steps between where you put the spaces and where the spaces end up that even if most of them mattered, it still wouldn't matter.